### DISCUSSION TITLE
SML_OBIS_LINE how does it work? ( + USE_SCRIPT_SUB_COMMAND )

### DISCUSSION THREAD

**Discussion by evlo** [arendst/Tasmota#16412] [Q&A]
I managed to compile 12.1.1.1 for esp32c3 and it finally works (12.0.* had bug with UART for esp32c3) anyway this is my script for parsing electrical meter
```
D
B
-sensor53 r
M 1
+1,20,o,0,300,Ele,21,200,2F3F210D0A
1,C.1.0(@1,SN,,sn,0
1,0.0.0(@1,Cust n.,,cn,0

;broken and will parse L1, L2, L3 into this as they also contain 1.8.0
;1,1.8.0(@1,Total in,KWh,Total_in,3
1,1.8.1(@1,T1 in,KWh,T1_in,3
1,1.8.2(@1,T2 in,KWh,T2_in,3

1,21.8.0(@1,L1 in,KWh,L1_in,3
1,41.8.0(@1,L2 in,KWh,L2_in,3
1,61.8.0(@1,L3 in,KWh,L3_in,3

1,2.8.0(@1,Total out,KWh,Total_out,3
;1,22.8.0(@1,L1 out,KWh,L1_out,3
;1,42.8.0(@1,L2 out,KWh,L2_out,3
;1,62.8.0(@1,L3 out,KWh,L3_out,3

;RRMMDDhhmm
;1,C.8.1(@1,T1 period,,t1_period,0
;1,C.8.2(@1,T2 period,,t2_period,0
;1,C.8.0(@1,Operating period in,,period_in,0
;1,C.82.0(@1,Operating period out,,period_out,0

;1,C.7.1(@1,Outages L1,,outages_L1,0
;1,C.7.2(@1,Outages L2,,outages_L2,0
;1,C.7.3(@1,Outages L3,,outages_L3,0

;1,0.2.1(@#44,SW,,sw_ver,0
;RRMMDDhhmm
;1,C.2.1(@1,last parametrisation,,last_parametrisation,0
;1,C.2.9(@1,last readout,,last_readout,0

;1,C.3.9(@1,Attacks,,attacks,0
#
```
i cn see in the console that tasmota does get data correctly
```
23:42:19.122 : 1.8.0(0016164.153*kWh)
23:42:19.125 : 21.8.0(0007631.425*kWh)
```
I had issue with parsing the data as 1.8.0 contains data and then it gets overwritten by 21.8.0 that contains different data.

So I added SML_OBIS_LINE to the defines and now tasmota only parses line `1,0.0.0(@1,Cust n.,,cn,0`
what did i do wrong?

I do also have USE_SCRIPT_SUB_COMMAND in the defs as i wanted to make readouts faster switching baudrate

**Comment by gemu2015** [comment]
normal obis parsing is done in shift mode, the comparator shifts the obis code char by char and compares with pattern.
this may fail if codes are contained in another code

SML_OBIS_LINE does not shift but reads line by line until a linefeed is detected, then compares that line
this should work with any legal obis code line.

USE_SCRIPT_SUB_COMMAND is not related to the SML driver

 you possibly mean USE_SML_SCRIPT_CMD which allows to send serial data to SML meters.

**Reply by evlo** [reply]
Hmmm, after removing SML_OBIS_LINE and USE_SCRIPT_SUB_COMMAND I do get the data as before.
What could be the reason for SML_OBIS_LINE just getting "0.0.0" line?

My meter sends data really like 21.8.0 and not in format of 1-0:1.8.1\*255 that is considered OBIS

I will try with just SML_OBIS_LINE later.

**Reply by gemu2015** [reply]
obis line delimiter should be CR LF. probably your meter has just CR then SML_OBIS_LINE would fail.
try this version which checks for that.
[xsns_53_sml.ino.zip](https://github.com/arendst/Tasmota/files/9476253/xsns_53_sml.ino.zip)
